File: Torre, Alfil y Rey/Torre, Alfil y Rey.py
#define Hd
def Rey(r1, c1, r2, c2):
    return max(abs(r1 - r2), abs(c1 - c2))
    
def Torre(r1, c1, r2, c2):
    if r1 == r2 or c1 == c2:
        return 1
    else:
        return 2
        
def Alfil(r1, c1, r2, c2):
    def puede(r1, c1, r2, c2):
        return (r1 + c1) %2 == abs(r2 - c2) %2
    if not puede(r1,c1,r2,c2):
        return 0
    elif abs(r1 - r2) == abs(c1 - c2):
        return 1
    else:
        return 2 
        
r1,c1,r2,c2=map(int,input().split())
print(Torre(r1,c1,r2,c2),Alfil(r1,c1,r2,c2),Rey(r1,c1,r2,c2))

File: Torre, Alfil y Rey/Torre, Alfil y Rey.pas
program Hd;

function Max(a,b:Integer):Integer; 
begin
  if a > b then Max:=a
  else Max:=b
end;

function Rey(r1, c1, r2, c2: Integer): Integer;
begin
  Rey := Max(Abs(r1 - r2), Abs(c1 - c2));
end;

function Torre(r1, c1, r2, c2: Integer): Integer;
begin
  if (r1 = r2) or (c1 = c2) then
    Torre := 1
  else
    Torre := 2;
end;

function puede(r1, c1, r2, c2: Integer): Boolean;
begin
  puede := (r1 + c1) mod 2 = Abs(r2 - c2) mod 2;
end;

function Alfil(r1, c1, r2, c2: Integer): Integer;
begin
  if not puede(r1, c1, r2, c2) then
    Alfil := 0
  else if Abs(r1 - r2) = Abs(c1 - c2) then
    Alfil := 1
  else
    Alfil := 2;
end;

var
  r1, c1, r2, c2: Integer;
begin
  ReadLn(r1, c1, r2, c2);
  WriteLn(Torre(r1, c1, r2, c2), ' ', Alfil(r1, c1, r2, c2), ' ', Rey(r1, c1, r2, c2));
end.